src/actions/languages/python/motion.ts

```diff
@@ -0,0 +1,232 @@
+import { BaseMovement } from '../../baseMotion';
+import { Position, TextDocument } from 'vscode';
+
+interface LineInfo {
+  line: number;
+  indentation: number;
+  text: string;
+}
+
+interface StructureElement {
+  type: 'function' | 'class';
+  start: Position;
+  end: Position;
+}
+
+/*
+ * Utility class used to parse the lines in the document and
+ * determine class and function boundaries
+ *
+ * The class keeps track of two positions: the ORIGINAL and the CURRENT
+ * using their relative locations to make decisions.
+ */
+export class PythonDocument {
+  _document: TextDocument;
+  structure: StructureElement[];
+
+  static readonly reOnlyWhitespace = /\S/;
+  static readonly reLastNonWhiteSpaceCharacter = /(?<=\S)\s*$/;
+  static readonly reDefOrClass = /^\s*(def|class) /;
+  static readonly reDef = /def/;
+
+  constructor(document: TextDocument) {
+    this._document = document;
+    const parsed = PythonDocument._parseLines(document);
+    this.structure = PythonDocument._parseStructure(parsed);
+  }
+
+  /*
+   * Generator of the lines of text in the document
+   */
+  static *lines(document: TextDocument): Generator<string> {
+    for (let index = 0; index < document.lineCount; index++) {
+      yield document.lineAt(index).text;
+    }
+  }
+
+  /*
+   * Calculate the indentation of a line of text.
+   * Lines consisting entirely of whitespace of "starting" with a comment are defined
+   * to have an indentation of "undefined".
+   */
+  static _indentation(line: string): number | undefined {
+    const index: number = line.search(PythonDocument.reOnlyWhitespace);
+
+    // Return undefined if line is empty, just whitespace, or starts with a comment
+    if (index === -1 || line[index] === '#') {
+      return undefined;
+    }
+
+    return index;
+  }
+
+  /*
+   * Parse a line of text to extract LineInfo
+   * Return undefined if the line is empty or starts with a comment
+   */
+  static _parseLine(index: number, text: string): LineInfo | undefined {
+    const indentation = this._indentation(text);
+
+    // Since indentation === 0 is a valid result we need to check for undefined explicitly
+    return indentation !== undefined ? { line: index, indentation, text } : undefined;
+  }
+
+  static _parseLines(document: TextDocument): LineInfo[] {
+    const lines = [...this.lines(document)]; // convert generator to Array
+    const infos = lines.map((text, index) => this._parseLine(index, text));
+
+    return infos.filter((x) => x) as LineInfo[]; // filter out empty/comment lines (undefined info)
+  }
+
+  static _parseStructure(lines: LineInfo[]): StructureElement[] {
+    const last = lines.length;
+    const structure: StructureElement[] = [];
+
+    for (let index = 0; index < last; index++) {
+      const info = lines[index];
+      const text = info.text;
+
+      if (text.match(PythonDocument.reDefOrClass)) {
+        const type = text.match(PythonDocument.reDef) ? 'function' : 'class';
```

Comment:
I think this distinction can be made by looking at the match group instead of running another regex, but probably not a big deal

Reply:
Note to self:

``` javascript
    const match = text.match(PythonDocument.reDefOrClass);
    if (match) {
        const type = match[1];
        ...
```

and delete the definition of `PythonDocument.reDef`.

Reply:
Done